File: acceptance/bin/fault.py
#!/usr/bin/env python3
"""Set up a fault rule on the testserver for the current test token.

Usage: fault.py [--after-handler] PATTERN STATUS_CODE OFFSET TIMES [ERROR_CODE]

  --after-handler  let the handler run and keep the state change it makes,
                   replacing only its response; models a response lost on the
                   way back from a request the backend already carried out
  PATTERN     HTTP method and path, supports trailing * wildcard,
              e.g. "PUT /api/2.0/permissions/pipelines/*"
  STATUS_CODE HTTP status code to return, e.g. 504
  OFFSET      number of requests to let through before fault starts
  TIMES       number of times to return the fault response
  ERROR_CODE  optional error_code for the response body, e.g.
              MAX_CHILD_NODE_SIZE_EXCEEDED (defaults to INJECTED)

The rule is scoped to the current DATABRICKS_TOKEN so it only affects
the test that registers it, even when tests share a server.
"""

import json
import os
import sys
import urllib.request

host = os.environ.get("DATABRICKS_HOST", "")
token = os.environ.get("DATABRICKS_TOKEN", "")

if not host:
    print("DATABRICKS_HOST not set", file=sys.stderr)
    sys.exit(1)

args = sys.argv[1:]
after_handler = "--after-handler" in args
if after_handler:
    args.remove("--after-handler")

if len(args) not in (4, 5):
    print(f"usage: {sys.argv[0]} [--after-handler] PATTERN STATUS_CODE OFFSET TIMES [ERROR_CODE]", file=sys.stderr)
    sys.exit(1)

pattern, status_code, offset, times = args[0], int(args[1]), int(args[2]), int(args[3])
error_code = args[4] if len(args) == 5 else "INJECTED"
body = json.dumps({"error_code": error_code, "message": "Fault injected by test."})

data = json.dumps(
    {
        "pattern": pattern,
        "status_code": status_code,
        "body": body,
        "offset": offset,
        "times": times,
        "after_handler": after_handler,
    }
).encode()

req = urllib.request.Request(
    f"{host}/__testserver/fault",
    data=data,
    headers={"Content-Type": "application/json", "Authorization": f"Bearer {token}"},
    method="POST",
)
urllib.request.urlopen(req)

File: libs/testserver/fault.go
package testserver

import (
	"bytes"
	"encoding/json"
	"strings"
	"sync"
)

type faultRuleKey struct {
	token   string
	pattern string
}

// FaultRule describes a single injected fault: HTTP status, body, and remaining fire count.
type FaultRule struct {
	StatusCode int
	Body       string
	// AfterHandler keeps the handler's state change and replaces only its response.
	AfterHandler bool
	offset       int
	times        int

	// bodyContains, if non-empty, additionally requires the request body to
	// contain this substring for the rule to fire. /workspace/import routes
	// every upload through the same method+path, so the only way to target a
	// single file's upload is by matching its multipart "path" form field.
	bodyContains string
}

// FaultRules holds the active fault injection rules for a test server.
type FaultRules struct {
	mu    sync.Mutex
	rules map[faultRuleKey]*FaultRule
}

// NewFaultRules returns an empty FaultRules.
func NewFaultRules() *FaultRules {
	return &FaultRules{rules: make(map[faultRuleKey]*FaultRule)}
}

// Set registers or replaces a fault rule for the given token and pattern.
func (fr *FaultRules) Set(token, pattern string, statusCode int, body string, offset, times int, bodyContains string) {
	fr.set(token, pattern, statusCode, body, offset, times, false, bodyContains)
}

// SetAfterHandler is like Set, but the handler runs first so its effect is kept.
func (fr *FaultRules) SetAfterHandler(token, pattern string, statusCode int, body string, offset, times int, bodyContains string) {
	fr.set(token, pattern, statusCode, body, offset, times, true, bodyContains)
}

func (fr *FaultRules) set(token, pattern string, statusCode int, body string, offset, times int, afterHandler bool, bodyContains string) {
	fr.mu.Lock()
	defer fr.mu.Unlock()
	fr.rules[faultRuleKey{token: token, pattern: pattern}] = &FaultRule{
		StatusCode:   statusCode,
		Body:         body,
		AfterHandler: afterHandler,
		offset:       offset,
		times:        times,
		bodyContains: bodyContains,
	}
}

// Check returns a matching fault rule and advances its counters, or nil if no rule matches.
// Pattern supports a trailing "*" wildcard, e.g. "PUT /api/2.0/permissions/pipelines/*".
// A rule with a non-empty bodyContains only matches when body contains that substring.
func (fr *FaultRules) Check(method, path, token string, body []byte) *FaultRule {
	requestPattern := method + " " + path

	fr.mu.Lock()
	defer fr.mu.Unlock()

	for key, rule := range fr.rules {
		if key.token != token {
			continue
		}
		rulePattern := key.pattern
		var matched bool
		if strings.HasSuffix(rulePattern, "*") {
			matched = strings.HasPrefix(requestPattern, rulePattern[:len(rulePattern)-1])
		} else {
			matched = requestPattern == rulePattern
		}
		if !matched {
			continue
		}
		// A non-matching body must not consume the rule's offset/times budget,
		// so this check precedes the counter updates below.
		if rule.bodyContains != "" && !bytes.Contains(body, []byte(rule.bodyContains)) {
			continue
		}
		if rule.offset > 0 {
			rule.offset--
			return nil
		}
		if rule.times <= 0 {
			delete(fr.rules, key)
			return nil
		}
		rule.times--
		if rule.times == 0 {
			delete(fr.rules, key)
		}
		result := *rule
		return &result
	}

	return nil
}

// faultEndpointHandler handles POST /__testserver/fault.
func faultEndpointHandler(fr *FaultRules) HandlerFunc {
	return func(req Request) any {
		var body struct {
			Pattern      string `json:"pattern"`
			StatusCode   int    `json:"status_code"`
			Body         string `json:"body"`
			Offset       int    `json:"offset"`
			Times        int    `json:"times"`
			AfterHandler bool   `json:"after_handler"`
			BodyContains string `json:"body_contains"`
		}
		if err := json.Unmarshal(req.Body, &body); err != nil {
			return Response{StatusCode: 400, Body: map[string]string{"error": err.Error()}}
		}
		if body.AfterHandler {
			fr.SetAfterHandler(req.Token, body.Pattern, body.StatusCode, body.Body, body.Offset, body.Times, body.BodyContains)
		} else {
			fr.Set(req.Token, body.Pattern, body.StatusCode, body.Body, body.Offset, body.Times, body.BodyContains)
		}
		return Response{StatusCode: 200}
	}
}
